tokio-executor/src/thread_pool/queue/local.rs

```diff
@@ -0,0 +1,308 @@
+use crate::task::Task;
+use crate::thread_pool::queue::global;
+
+use std::fmt;
+use std::mem::MaybeUninit;
+use std::ptr;
+
+pub(super) struct Queue<T: 'static> {
+    /// Concurrently updated by many threads.
+    head: AtomicU32,
+
+    /// Only updated by producer thread but read by many threads.
+    tail: AtomicU32,
+
+    mask: usize,
+
+    /// Elements
+    buffer: Box<[CausalCell<MaybeUninit<Task<T>>>]>,
+}
+
+impl<T: 'static> Queue<T> {
+    pub(super) fn with_capacity(capacity: usize) -> Queue<T> {
+        assert!(capacity >= 2 && capacity.is_power_of_two());
+
+        let mut buffer = Vec::with_capacity(capacity);
+
+        for _ in 0..capacity {
+            buffer.push(CausalCell::new(MaybeUninit::uninit()));
+        }
+
+        Queue {
+            head: AtomicU32::new(0),
+            tail: AtomicU32::new(0),
+            mask: capacity - 1,
+            buffer: buffer.into(),
+        }
+    }
+}
+
+impl<T> Queue<T> {
+    /// Push a task onto the local queue.
+    ///
+    /// This **must** be called by the producer thread.
+    pub(super) unsafe fn push(&self, mut task: Task<T>, global: &global::Queue<T>) {
+        loop {
+            let head = self.head.load(Acquire);
+
+            // safety: this is the **only** thread that updates this cell.
+            let tail = self.tail.unsync_load();
+
+            if tail.wrapping_sub(head) < self.buffer.len() as u32 {
+                // Map the position to a slot index.
+                let idx = tail as usize & self.mask;
+
+                self.buffer[idx].with_mut(|ptr| {
+                    // Write the task to the slot
+                    ptr::write((*ptr).as_mut_ptr(), task);
+                });
+
+                // Make the task available
+                self.tail.store(tail + 1, Release);
+
+                return;
+            }
+
+            // The local buffer is full. Push a batch of work to the global
+            // queue.
+            match self.push_overflow(task, head, tail, global) {
+                Ok(_) => return,
+                // Lost the race, try again
+                Err(v) => task = v,
+            }
+
+            atomic::spin_loop_hint();
+        }
+    }
+
+    /// Move a batch of tasks into the global queue.
+    ///
+    /// This will temporarily make some of the tasks unavailable to stealers.
+    /// Once `push_overflow` is done, a notification is sent out, so if other
+    /// workers "missed" some of the tasks during a steal, they will get
+    /// another opportunity.
+    #[inline(never)]
+    unsafe fn push_overflow(
+        &self,
+        task: Task<T>,
+        head: u32,
+        tail: u32,
+        global: &global::Queue<T>,
+    ) -> Result<(), Task<T>> {
+        use super::LOCAL_QUEUE_CAPACITY;
+
+        const BATCH_LEN: usize = LOCAL_QUEUE_CAPACITY / 2 + 1;
+
+        let n = tail.wrapping_sub(head) / 2;
+        assert_eq!(n as usize, LOCAL_QUEUE_CAPACITY / 2, "queue is not full");
```

Comment:
Should this use `self.buffer.len()` instead of `LOCAL_QUEUE_CAPACITY`?

Reply:
Good call. The length of the queue is hard coded, but we should not be using both.